### PR TITLE
New Template : Red Hat Linux 3 Tier IaaS on Azure

### PR DESCRIPTION
**Pull request by amalik99**
### Best Practice Checklist
Check these items before submitting a PR... See the Contribution Guide for the full detail: https://github.com/Azure/azure-quickstart-templates/blob/master/1-CONTRIBUTION-GUIDE/README.md 

1. uri's compatible with all clouds (Stack, China, Government)
1. Staged artifacts use _artifactsLocation & _artifactsLocationSasToken
1. Use resourceGroup().location for resource locations
1. Folder names for artifacts (nestedtemplates, scripts, DSC)
1. Use literal values for apiVersion (no variables)
1. Parameter files (GEN-UNIQUE for value generation and no "changemeplease" values
1. $schema and other uris use https
1. Use uniqueString() whenever possible to generate names for resources.  While this is not required, it's one of the most common failure points in a deployment. 
1. Update the metadata.json with the current date

For details: https://github.com/Azure/azure-quickstart-templates/blob/master/1-CONTRIBUTION-GUIDE/bp-checklist.md

- [x] - Please check this box once you've submitted the PR if you've read through the Contribution Guide and best practices checklist.

### Changelog

*
*
*

### CODE REVIEW

**Comment by azurecla** [comment]
Hi __@amalik99__, I'm your friendly neighborhood Azure Pull Request Bot (You can call me AZPRBOT). Thanks for your contribution!
    <span>You've already signed the contribution license agreement. Thanks!</span>
        <span>We will now validate the agreement and then real humans will evaluate your PR.</span>

TTYL, AZPRBOT;

File: rhel-3tier-iaas/README.md [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+![Deployment Solution Architecture](https://github.com/Azure/azure-quickstart-templates/master/redhat-3tier-architecture/images/architecture.png?raw=true)
+
+##Licenses and Costs 
+
+The Red Hat Enterprise Linux 7.3 image used in this solution is the PAYG model and doesn't require the user to license it, it will be licensed automatically after the instance is launched first time. Use of this image carries a separate hourly charge that is in addition to Microsoft's Linux VM rates. Total price of the VM consists of the base Linux VM price plus RHEL VM image surcharge.  Click [here](https://azure.microsoft.com/en-us/pricing/details/virtual-machines/red-hat/) for pricing details.
+
+##Prerequisites 
+
+- Azure Subscription with specified payment method (Red Hat Enterprise Linux is a market place product and requires payment method to be specified in Azure Subscription
+-	User account with admin or contributor access to the given azure subscription
```
Comment (bmoore-msft) [inline rhel-3tier-iaas/README.md]: Is this prereq needed only for template deployment or something else?  (all samples in the repo need contributor access)

File: rhel-3tier-iaas/azuredeploy.json [hunk 2]
```diff
@@ -0,0 +1,951 @@
...
+      "databaseTierVmCount":{
+         "type":"int",
+         "defaultValue":2,
+         "metadata":{
+            "description":"Number of Database servers to be deployed"
+         }
+      }
+   },
+   "variables":{
+      "webTierVmSize":"Standard_D2_v2",
```
Comment (bmoore-msft) [inline rhel-3tier-iaas/azuredeploy.json]: There are a number of variables that are only used once - can we just use literal property values for those?

File: rhel-3tier-iaas/azuredeploy.json [hunk 3]
```diff
@@ -0,0 +1,951 @@
...
+      "jumpSubnetName":"jump-subnet",
+      "webNSGName":"web-tier-NSG",
+      "appNSGName":"app-tier-NSG",
+      "databaseNSGName":"database-tier-NSG",
+      "jumpNSGName":"jump-NSG",
+      "webLoadBalancerName":"web-lb",
+      "weblbIPAddressName":"web-lb-pip",
+      "weblbDnsLabel":"[concat('weblb',uniqueString(resourceGroup().id))]",
+      "webLoadBalancerIPID":"[resourceId('Microsoft.Network/publicIPAddresses',variables('weblbIPAddressName'))]",
+      "webFrontEndIPConfigID":"[concat(resourceId('Microsoft.Network/loadBalancers',variables('webLoadBalancerName')),'/frontendIPConfigurations/loadBalancerFrontEnd')]",
```
Comment (bmoore-msft) [inline rhel-3tier-iaas/azuredeploy.json]: just use resourceId() for the multi-part resourceId()s (here and below) - get rid of the unnecessary concat()

File: rhel-3tier-iaas/azuredeploy.json [hunk 4]
```diff
@@ -0,0 +1,951 @@
...
+      "appTierVmName":"app-tier-vm",
+      "appTierOSDiskName":"appTierOSDisk",
+      "databaseTierVmName":"database-tier-vm",
+      "databaseTierOSDiskName":"dbTierOSDisk",
+      "jumpVmName":"jump-vm",
+      "jumpOSDiskName":"jumpVmOSDisk"
+   },
+   "resources":[
+      {
+         "type":"Microsoft.Storage/storageAccounts",
```
Comment (bmoore-msft) [inline rhel-3tier-iaas/azuredeploy.json]: using a copy loop for these resources would remove quite a few lines of code...

File: rhel-3tier-iaas/azuredeploy.json [hunk 5]
```diff
@@ -0,0 +1,951 @@
...
+         "name":"[variables('appNSGName')]",
+         "apiVersion":"2016-03-30",
+         "location":"[resourceGroup().location]",
+         "comments":"NSG for App Subnet",
+         "tags":{
+            "displayName":"App NSG"
+         },
+         "properties":{
+            "securityRules":[
+
```
Comment (bmoore-msft) [inline rhel-3tier-iaas/azuredeploy.json]: add something to the comment property to explain why these are empty

File: rhel-3tier-iaas/azuredeploy.json [hunk 6]
```diff
@@ -0,0 +1,951 @@
...
+         "name":"[variables('databaseNSGName')]",
+         "apiVersion":"2016-03-30",
+         "location":"[resourceGroup().location]",
+         "comments":"NSG for Database Subnet",
+         "tags":{
+            "displayName":"Database NSG"
+         },
+         "properties":{
+            "securityRules":[
+
```
Comment (bmoore-msft) [inline rhel-3tier-iaas/azuredeploy.json]: same as above

File: rhel-3tier-iaas/azuredeploy.json [hunk 7]
```diff
@@ -0,0 +1,951 @@
...
+                     "networkSecurityGroup":{
+                        "id":"[resourceId('Microsoft.Network/networkSecurityGroups', variables('jumpNSGName'))]"
+                     }
+                  }
+               }
+            ]
+         }
+      },
+      {
+         "type":"Microsoft.Compute/availabilitySets",
```
Comment (bmoore-msft) [inline rhel-3tier-iaas/azuredeploy.json]: consider a copy loop for the avail sets as well

File: rhel-3tier-iaas/azuredeploy.json [hunk 8]
```diff
@@ -0,0 +1,951 @@
...
+            "count":"[parameters('webTierVmCount')]"
+         },
+         "properties":{
+            "ipConfigurations":[
+               {
+                  "name":"ipconfig1",
+                  "properties":{
+                     "privateIPAllocationMethod":"Dynamic",
+                     "subnet":{
+                        "id":"[concat(resourceId('Microsoft.Network/virtualNetworks', variables('virtualNetworkName')), '/subnets/', variables('webTierSubnetName'))]"
```
Comment (bmoore-msft) [inline rhel-3tier-iaas/azuredeploy.json]: just use resourceId()

File: rhel-3tier-iaas/azuredeploy.json [hunk 9]
```diff
@@ -0,0 +1,951 @@
...
+            "count":"[parameters('appTierVmCount')]"
+         },
+         "properties":{
+            "ipConfigurations":[
+               {
+                  "name":"ipconfig1",
+                  "properties":{
+                     "privateIPAllocationMethod":"Dynamic",
+                     "subnet":{
+                        "id":"[concat(resourceId('Microsoft.Network/virtualNetworks', variables('virtualNetworkName')), '/subnets/', variables('appTierSubnetName'))]"
```
Comment (bmoore-msft) [inline rhel-3tier-iaas/azuredeploy.json]: just use resourceId()

File: rhel-3tier-iaas/azuredeploy.json [hunk 10]
```diff
@@ -0,0 +1,951 @@
...
+            "count":"[parameters('databaseTierVmCount')]"
+         },
+         "properties":{
+            "ipConfigurations":[
+               {
+                  "name":"ipconfig1",
+                  "properties":{
+                     "privateIPAllocationMethod":"Dynamic",
+                     "subnet":{
+                        "id":"[concat(resourceId('Microsoft.Network/virtualNetworks', variables('virtualNetworkName')), '/subnets/', variables('databaseTierSubnetName'))]"
```
Comment (bmoore-msft) [inline rhel-3tier-iaas/azuredeploy.json]: just use resourceId()

File: rhel-3tier-iaas/azuredeploy.json [hunk 11]
```diff
@@ -0,0 +1,951 @@
...
+            "ipConfigurations":[
+               {
+                  "name":"ipconfig1",
+                  "properties":{
+                     "privateIPAllocationMethod":"Dynamic",
+                     "publicIPAddress":{
+                        "id":"[resourceId('Microsoft.Network/publicIPAddresses',variables('jumpIPAddressName'))]"
+                     },
+                     "subnet":{
+                        "id":"[concat(resourceId('Microsoft.Network/virtualNetworks', variables('virtualNetworkName')), '/subnets/', variables('jumpSubnetName'))]"
```
Comment (bmoore-msft) [inline rhel-3tier-iaas/azuredeploy.json]: just use resourceId()

File: rhel-3tier-iaas/README.md (hunk 1 above)
Comment (amalik99) [inline rhel-3tier-iaas/README.md]: only for template deployment, will remove this as its kinda pre-req for all template in repo.

File: rhel-3tier-iaas/azuredeploy.json (hunk 2 above)
Comment (amalik99) [inline rhel-3tier-iaas/azuredeploy.json]: Converted into single variable named vmSize

**Comment by amalik99** [comment]
Hi Brian, Thanks for the review. All requested changes have been implemented, creating Diagnostics storage account outside copy loop to maintain its naming differentiation.

Let me know in case of any other suggestions.

**Pull request merged**